src/neoxp/Commands/ContractCommand.Download.cs

```diff
@@ -0,0 +1,75 @@
+using System;
+using System.ComponentModel.DataAnnotations;
+using System.IO;
+using System.Net.Http;
+using System.Threading.Tasks;
+using McMaster.Extensions.CommandLineUtils;
+using Neo;
+using Neo.IO.Json;
+using Neo.Network.RPC;
+
+namespace NeoExpress.Commands
+{
+    partial class ContractCommand
+    {
+        [Command(Name = "download", Description = "Download contract with storage from remote chain into local chain")]
+        internal class Download
+        {
+            readonly ExpressChainManagerFactory chainManagerFactory;
+            private RpcClient rpcClient = null!;
+            
+            public Download(ExpressChainManagerFactory chainManagerFactory)
+            {
+                this.chainManagerFactory = chainManagerFactory;
+            }
+            
+            [Argument(0, Description = "Contract invocation hash")]
+            [Required]
+            internal string Contract { get; init; } = string.Empty;
+            
+            [Argument(1, Description = "Source network RPC address")]
+            internal string Source { get; init; } = string.Empty;
+            
+            [Option(Description = "Path to neo-express data file")]
+            internal string Input { get; init; } = string.Empty;
+
+            internal async Task ExecuteAsync(TextWriter writer)
```

Comment:
Other commands break `ExecuteAsync` out into a separate method so it can be called as part of the `BatchCommand`. I'm thinking `download` should also be supported by batch.

Reply:
Sorry I don't understand what's expected. I used `ContractCommand.Get` as base to implement the `Download` command. Looking at other `Contract` commands like `Deploy`, `Hash` and `Invoke` all I see is that they only implement `OnExecuteAsync` and not `ExecuteAsync`. The distinction between the 2 and their use-cases is not clear to me which is probably why I don't understand the request. If you can elaborate on this I can probably fix the code. Thanks

Reply:
My bad, I wasn't clear and my decription of the code was wrong.

Operations that modify the state of the blockchain are typically also supported by the `batch` command. Older versions of express would put the common code used by both direct and `batch` versions of the command into `ExecuteAsync`, but the `TransactionExecutor` class handles most of that now. Obviously, `download` modifications are not transactions, so `TransactionExecutor` would not be a good place to put `download` code shared between direct and `batch` versions.

I actually need to think about this more. THe only other command that connects to a public blockchain is `policy sync`, but the `batch` version of that command *explicitly does not support* downloading the policy from the public network when the `batch` command is run. I feel strongly that the `batch` command must be deterministic, so the `batch` version of `policy sync` only supports syncing with a previously downloaded policy state JSON file. By that logic, `batch` version of `download` should NOT support getting whatever the current version of contract + state is currently live on the public network. However, since we're using `StateService` anyway, we could force the `batch` version of `download` to specify a block height, ensuring deterministic update of the local express environment. 

Maybe it would be easier to discuss directly on Discord?